src/events.rs

```diff
@@ -1,26 +1,92 @@
 use crate::auth::{Authorization, AuthorizationError, Capabilities};
 use crate::config::Config;
 use crate::publish::{publish, Sequencing, MESSAGE_SIZE_MAX};
-use crate::storage::Storage;
+use crate::storage::{RetainedVersion, Storage, StorageError};
+use base64::Engine;
 use fastly::http::{header, StatusCode};
 use fastly::{Request, Response};
-use std::collections::{HashMap, HashSet};
+use std::collections::HashMap;
+use std::fmt::Write;
 use std::str;
 use std::time::Duration;
+use thiserror::Error;
 
 const TOPICS_PER_REQUEST_MAX: usize = 10;
 
+struct VersionParseError;
+
+#[derive(Debug, Copy, Clone)]
 struct Version {
     generation: u64,
     seq: u64,
 }
 
 impl Version {
-    fn to_id(&self) -> String {
-        format!("{}-{}", self.generation, self.seq)
+    fn as_id(&self) -> String {
+        format!("{:16x}-{}", self.generation, self.seq)
+    }
+
+    fn parse(s: &str) -> Result<Self, VersionParseError> {
+        let pos = match s.find('-') {
+            Some(pos) => pos,
+            None => return Err(VersionParseError),
+        };
+
+        let generation = &s[..pos];
+        let seq = &s[(pos + 1)..];
```

Comment:
```suggestion
        let generation = &s.get(..pos).ok_or(VersionParseError)?;
        let seq = &s.get((pos + 1)..).ok_or(VersionParseError)?;
```

Substring slicing always gives me pause, would something like this be a little better?